### ISSUE TITLE
Russian unicode char is irretrievably lost in Ctrl+RussianLetter key events

### ISSUE THREAD

**Issue by unxed** [wxWidgets/wxWidgets#23477]
1. Set two keyboard layouts system wide: English one and Russian one
2. Open Keyboard sample app form wx
3. Switch to Russian layout
4. Press `Ctrl`+`я` (same physical keys as `Ctrl`+`z`)
5. You will see `Ctrl-Z` (26) value in both keycode and unicode key field. So letter `я` is lost forever.

This reproduces in both GTK and Windows backends.

My suggestion: we should put original unicode letter (like `я` in this case) in unicode key field and leave current value in keycode field only.

This can break backward compatibility, but we are breaking it a bit anyway with recent GTK backend keyboard events behavior changes. Also special ctrl+letters parsing is aimed to help in terminal apps development, but the only somewhat popular wx-based terminal app I know is far2l, and we will make necessary adoptions on our side if needed.

**Comment by vadz** [comment]
Yes, I think we could change the Unicode key for such events, all the existing code almost surely only looks at the key code for them anyhow, so as long as this one remains as it is now, i.e. 26, it should be fine.

**Comment by unxed** [comment]
Can write such PR for GTK, but afraid to change anything other backends as I do not have Windows or Mac for testing.

**Comment by vadz** [comment]
I could test it under both.

**Comment by unxed** [comment]
For now I can't even find any analog of AdjustCharEventKeyCodes() in MSW backend. Maybe such processing is done by Windows itself?

For GTK, changes are simple, like
```
diff --git a/src/gtk/window.cpp b/src/gtk/window.cpp
--- a/src/gtk/window.cpp
+++ b/src/gtk/window.cpp
@@ -1290,7 +1290,6 @@ bool AdjustCharEventKeyCodes(wxKeyEvent& event)
         // Adjust the Unicode equivalent in the same way too.
         if ( event.m_keyCode != code )
         {
-            event.m_uniChar = event.m_keyCode;
             modified = true;
         }
     }
@@ -1429,6 +1428,9 @@ gtk_window_key_press_callback( GtkWidget *WXUNUSED(widget),
                 // use Unicode values
                 eventChar.m_keyCode = key_code;
                 eventChar.m_uniChar = uniChar;
+            } else if (event.ControlDown()) {
+                // for Ctrl+Letters use unicode values also
+                eventChar.m_uniChar = uniChar;
             }
 
             wxLogTrace(TRACE_KEYS, wxT("Char event: %ld"), eventChar.m_keyCode);

```

**Comment by unxed** [comment]
Also, as I can see looking on OSX code, this issue should not affect it as translations are done only for keycode field and I see not direct assignments of keycode field value to unicodekey field. But it's better to verify it.

**Comment by vadz** [comment]
> For now I can't even find any analog of AdjustCharEventKeyCodes() in MSW backend. Maybe such processing is done by Windows itself?

Yes, I think so.

> Also, as I can see looking on OSX code, this issue should not affect it as translations are done only for keycode field and I see not direct assignments of keycode field value to unicodekey field. But it's better to verify it.

It does affect it, Unicode key corresponds to a Latin letter for such events there too.

Funnily enough, when I press "Cmd" symbol on the virtual keyboard using Russian layout, the letters on it change to Latin, i.e. the OS virtual keyboard seems to consider that Cmd-letter does actually generate Latin letters, see the initial state:

![initial](https://user-images.githubusercontent.com/146917/233851699-b3542027-3b71-4394-b469-f61b67102b30.png)

and the state after pressing "Cmd" (it's highlighted to indicate that it's active):

![after_cmd](https://user-images.githubusercontent.com/146917/233851726-0da5c526-d823-4b2f-9037-32c92a1df2dc.png)

**Comment by unxed** [comment]
Made some testing with Wine. It looks like normal Windows behavior is to send 1..26 for Ctrl+letters independently of keyboard layout selected. But it's still possible to get corresponding unicode char by using
```
ToUnicode(wParam, HIWORD(lParam) & 0xFF, keyboardState, unicodeChar, 1, 0);
```
if keyboardState is set to zero, ignoring Control key state (actially it should still contain real Shift state, not all zeros, to maintain correct character case, but I am just experimenting for now).

**Comment by unxed** [comment]
For WM_CHAR events, getting unicode key value for Ctrl+letter looks like
```
        memset(keyboardState, 0, sizeof(keyboardState));
        int vkc = MapVirtualKey(HIWORD(lParam) & 0xFF, MAPVK_VSC_TO_VK);
        result = ToUnicode(vkc, HIWORD(lParam) & 0xFF, keyboardState, unicodeChar, 1, 0);
        printf("Unicode char w/o modifiers: %i\n", unicodeChar[0]);
```

it's just proof of concept, Shift state is also ignored here (but it should not be)

**Comment by unxed** [comment]
Not sure is we can do anything with Macs, btw.

**Comment by unxed** [comment]
Small PoC app showing how to find Unicode Char value for Ctrl+non_latin_letter events on Windows:
```
#define UNICODE
#define _UNICODE

#include <windows.h>
#include <stdio.h>

LRESULT CALLBACK WindowProc(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam);

int CALLBACK WinMain(_In_ HINSTANCE hInstance, _In_opt_ HINSTANCE hPrevInstance, _In_ LPSTR lpCmdLine, _In_ int nCmdShow)
{
    const wchar_t CLASS_NAME[] = L"Simple Key Logger";

    WNDCLASS wc = {};
    wc.lpfnWndProc = WindowProc;
    wc.hInstance = hInstance;
    wc.lpszClassName = CLASS_NAME;

    RegisterClass(&wc);

    HWND hwnd = CreateWindowEx(
        0,
        CLASS_NAME,
        L"Simple Key Logger",
        WS_OVERLAPPEDWINDOW,
        CW_USEDEFAULT, CW_USEDEFAULT, CW_USEDEFAULT, CW_USEDEFAULT,
        NULL,
        NULL,
        hInstance,
        NULL
    );

    if (hwnd == NULL)
    {
        return 0;
    }

    ShowWindow(hwnd, nCmdShow);

    MSG msg = {};
    while (GetMessage(&msg, NULL, 0, 0))
    {
        TranslateMessage(&msg);
        DispatchMessage(&msg);
    }

    return 0;
}

LRESULT CALLBACK WindowProc(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
{
    BYTE keyboardState[256];
    int result;
    wchar_t unicodeChar[2] = {0};

    switch (uMsg)
    {
    case WM_DESTROY:
        PostQuitMessage(0);
        return 0;

    case WM_KEYDOWN:
        printf("WM_KEYDOWN, key code: %i\n", wParam);

        GetKeyboardState(keyboardState);
        result = ToUnicode(wParam, HIWORD(lParam) & 0xFF, keyboardState, unicodeChar, 1, 0);
        printf("Unicode char: %i\n", unicodeChar[0]);

        memset(keyboardState, 0, sizeof(keyboardState));
        result = ToUnicode(wParam, HIWORD(lParam) & 0xFF, keyboardState, unicodeChar, 1, 0);
        printf("Unicode char w/o modifiers: %i\n", unicodeChar[0]);

        printf("\n");
        return 0;

    case WM_UNICHAR:
        printf("WM_UNICHAR, char code: %i\n", wParam);
        printf("\n");
        return 0;

    case WM_CHAR:
        printf("WM_CHAR, char code: %i\n", wParam);

        memset(keyboardState, 0, sizeof(keyboardState));
        int vkc = MapVirtualKey(HIWORD(lParam) & 0xFF, MAPVK_VSC_TO_VK);
        result = ToUnicode(vkc, HIWORD(lParam) & 0xFF, keyboardState, unicodeChar, 1, 0);
        printf("Unicode char w/o modifiers: %i\n", unicodeChar[0]);
        
        printf("\n");
        return 0;

    default:
        return DefWindowProc(hwnd, uMsg, wParam, lParam);
    }
}

```

**Comment by unxed** [comment]
Is where any way to cross compile windows wx version by using something like i686-w64-mingw32-gcc on Linux?

**Comment by vadz** [comment]
> Is where any way to cross compile windows wx version by using something like i686-w64-mingw32-gcc on Linux?

Yes, of course, this works out of the box, just use `--host=i686-w64-mingw32` with configure. See https://github.com/wxWidgets/wxWidgets/blob/master/.github/workflows/ci_msw_cross.yml

**Comment by unxed** [comment]
Implemented for GTK and MSW in #23494, please review. Still no idea what to do with MacOS.

**Comment by unxed** [comment]
I see some other strange things with Ctrl+CharKeys.

On MSW, Ctrl + non-letter keys as `.` or `;` or numbers produce no OnChar events at all.
Ctrl+`[`, `]`, `\` produces 27, 29, 28 keycodes.
Ctrl+Shift+`-` produces 31 keycode.

This all probably needs investigating and syncing behavior across different platforms.

**Comment by unxed** [comment]
More of this, Ctrl + numpad8 generates DELETE keycode in onchar if numlock is on! Other numpad keys also behave strange with Ctrl.

Shift also affects numpad onchar keycodes. For example, with numlock on shift+numpad8 gives UP keycode instead of 8.
